Refactor certain doctests towards either reStructuredText documentation or Python tests (#464)
crate/crate-python: docs/by-example/sqlalchemy/advanced-querying.rst

.. _sqlalchemy-advanced-querying:

=============================
SQLAlchemy: Advanced querying
=============================

This section of the documentation demonstrates running queries with
aggregations, and queries using a fulltext index with analyzer, both
using the CrateDB SQLAlchemy dialect.

.. rubric:: Table of Contents

.. contents::
   :local:

Introduction
============

Import the relevant symbols:

    >>> import sqlalchemy as sa
    >>> from sqlalchemy.ext.declarative import declarative_base
    >>> from sqlalchemy.orm import sessionmaker
    >>> from uuid import uuid4

Establish a connection to the database, see also :ref:`sa:engines_toplevel`
and :ref:`connect`:

    >>> engine = sa.create_engine(f"crate://{crate_host}")
    >>> connection = engine.connect()

Create an SQLAlchemy :doc:`Session <sa:orm/session_basics>`:

    >>> session = sessionmaker(bind=engine)()
    >>> Base = declarative_base()

Introduction to fulltext indexes
================================

> :ref:`crate-reference:fulltext-indices` take the contents of one or more fields
> and split it up into tokens that are used for fulltext-search. The transformation
> from a text to separate tokens is done by an analyzer. In order to create
> fulltext search queries a :ref:`fulltext index with an analyzer
> <crate-reference:sql_ddl_index_fulltext>` must be defined for the related columns.

Reviewer: Could restructure this a bit to make the flow more natural. For example:

```suggestion
:ref:`crate-reference:fulltext-indices` take the contents of one or more fields
and split it up into tokens that are used for fulltext-search. The
transformation from a text to separate tokens is done by an analyzer. In order
to create fulltext search queries we need to create a table with a
:ref:`fulltext index with an analyzer
<crate-reference:sql_ddl_index_fulltext>`::

    CREATE TABLE characters (
        id STRING PRIMARY KEY,
        name STRING,
        quote STRING,
        INDEX name_ft USING fulltext(name) WITH (analyzer = 'english'),
        INDEX quote_ft USING fulltext(quote) WITH (analyzer = 'english')
    )

We have to create this table using SQL because it is currently not possible to create ``INDEX`` fields using SQLAlchemy's declarative mapping. But we can define the table to use all other operations:

    >>> def gen_key():
    ...     return str(uuid4())

    >>> class Character(Base):
    ...     __tablename__ = 'characters'
    ...     id = sa.Column(sa.String, primary_key=True, default=gen_key)
    ...     name = sa.Column(sa.String)
    ...     quote = sa.Column(sa.String)
    ...     name_ft = sa.Column(sa.String)
    ...     quote_ft = sa.Column(sa.String)
    ...     __mapper_args__ = {
    ...         'exclude_properties': ['name_ft', 'quote_ft']
    ...     }

We define ``name_ft`` and ``quote_ft`` as regular columns but add them under ``__mapper_args__.exclude_properties`` to ensure they're excluded from insert or update operations.
```

.. code-block:: sql

    CREATE TABLE characters (
        id STRING PRIMARY KEY,
        name STRING,
        quote STRING,
        INDEX name_ft USING fulltext(name) WITH (analyzer = 'english'),
        INDEX quote_ft USING fulltext(quote) WITH (analyzer = 'english')
    )

.. note::

    Currently, it is not supported to define a *named index column definition*
    using a :ref:`crate-reference:sql_ddl_index_fulltext` on behalf of the
    SQLAlchemy declarative schema.

Let's add two records we use for testing.

    >>> arthur = Character(name='Arthur Dent')
    >>> arthur.quote = "Let's go somewhere."
    >>> session.add(arthur)

    >>> trillian = Character(name='Tricia McMillan')
    >>> trillian.quote = "We're on a space ship Arthur. In space."
    >>> session.add(trillian)

    >>> session.commit()

After ``INSERT`` statements are submitted to the database, the newly inserted
records aren't immediately available for retrieval because the index is only
updated periodically (default: each second). In order to synchronize that,
refresh the table:

    >>> _ = connection.execute(sa.text("REFRESH TABLE characters"))

Aggregates: Counting and grouping

> =================================

Reviewer: Maybe move this elsewhere - the previous section is "Introduction to fulltext indexes" and the next section is "Fulltext search with MATCH predicate" - having an intermediate aggregation/count/group-by breaks the flow

Author: I agree and will try to find a different spot.

Author: The section has been moved one slot to the bottom with fe9030016.

SQLAlchemy supports different approaches to issue a query with a count
aggregate function. Take a look at the `count result rows`_ documentation
for a full overview.

CrateDB currently does not support all variants as it can not handle the

> sub-queries yet.

Reviewer: Is this still the case?

Author: Good question. I also wanted to highlight this note while reading it, but missed to do so. Which suggestion do you have?
a) Investigate the matter.
b) Just remove this sentence.

Author: I will do. However, to move on with this, and other subsequent patches, I've deferred it to GH-489. I hope this is fine.

This means that queries using ``count()`` have to be written in one of the
following ways:

    >>> session.query(sa.func.count(Character.id)).scalar()
    2

    >>> session.query(sa.func.count('*')).select_from(Character).scalar()
    2

Using the ``group_by`` clause is similar:

    >>> session.query(sa.func.count(Character.id), Character.name) \
    ...     .group_by(Character.name) \
    ...     .order_by(sa.desc(sa.func.count(Character.id))) \
    ...     .order_by(Character.name).all()
    [(1, 'Arthur Dent'), (1, 'Tricia McMillan')]

Fulltext search with MATCH predicate
====================================

> Fulltext search in CrateDB is performed using the :ref:`crate-reference:predicates_match`.

Reviewer: ```suggestion
Fulltext search in CrateDB is performed using :ref:`crate-reference:predicates_match`.
```

The CrateDB SQLAlchemy dialect comes with a ``match`` function, which can be used to
search on one or multiple fields.

    >>> from crate.client.sqlalchemy.predicates import match

    >>> session.query(Character.name) \
    ...     .filter(match(Character.name_ft, 'Arthur')) \
    ...     .all()
    [('Arthur Dent',)]

> To get the relevance of a matching row, an internal system column ``_score``
> can be selected. It is a numeric value which is relative to the other rows.

Reviewer: ```suggestion
To get the relevance of a matching row, you can select the ``_score`` system
column. It is a numeric value which is relative to the other rows.
```

The higher the score value, the more relevant the row.

In most cases, ``_score`` is not part of the SQLAlchemy table definition,
so it must be passed as a string:

    >>> session.query(Character.name, sa.literal_column('_score')) \
    ...     .filter(match(Character.quote_ft, 'space')) \
    ...     .all()
    [('Tricia McMillan', ...)]

> To search on multiple columns, pass a dictionary with columns and ``boost``
> attached. ``boost`` is a factor that increases the relevance of a column in
> respect to the other columns:

Reviewer: ```suggestion
To search multiple columns, use a dictionary where the keys are the columns and the values are a ``boost``. A ``boost`` is a factor that increases the relevance of a column in respect to the other columns:
```

    >>> session.query(Character.name) \
    ...           .filter(match({Character.name_ft: 1.5, Character.quote_ft: 0.1},
    ...                         'Arthur')) \
    ...     .order_by(sa.desc(sa.literal_column('_score'))) \
    ...     .all()
    [('Arthur Dent',), ('Tricia McMillan',)]

The ``match_type`` argument determines how a single ``query_term`` is applied
and how the resulting ``_score`` is computed. Thus, it influences which
documents are considered more relevant. The default selection is ``best_fields``.
For more information, see :ref:`crate-reference:predicates_match_types`.

Results are ordered by ``_score`` by default, but can be overridden by adding

> an ``order_by()`` clause.

Reviewer: Is this really the case? Afaik we don't have a default ordering for them (or at least we don't ensure that it's kept when merging results from multiple nodes)

Author: All right. Then let's just remove this sentence? Or should we clarify "the other way round", like:

```suggestion
Because merging results from multiple CrateDB nodes does not yield sorted records,
you would have to add an ``order_by()`` clause in order to sort by ``_score``.
```

Reviewer: I'd just mention that if you want to have the results sorted by the score you can use the `order_by` function.

Without any elaboration regarding default ordering.

Author: Proposal:
```suggestion
Results are not sorted by default. If you want to sort them by ``_score``, you
can use the ``order_by()`` function.
```
